ksqldb-engine/src/main/java/io/confluent/ksql/engine/EngineExecutor.java

```diff
@@ -164,10 +164,12 @@ PullQueryResult executePullQuery(
           ksqlConfig,
           analysis,
           statement);
-      final HARouting routing = new HARouting(
+
+      try (HARouting routing = new HARouting(
```

Comment:
This must have accidentally been introduced in the refactor.  I'm surprised it didn't show up in benchmarking if the threads were being leaked.  I think that technically, it would have been eventually garbage collected once the executor was GCed, but it is certainly preferable to shut it down immediately.

Reply:
Are we sure the change was properly picked up when benchmarked? It looks like it used to be created once per `PullQueryExecutor` (which I think was once per request). It could be that the benchmarks didn't catch it because the `newFixedThreadPool` is good about not allocating 100 threads unless it needs them (so we were really only leaking one thread each time). I still suspect that this would have very quickly caused OOM (cannot allocate native thread) after we issue a few thousand pull queries.

Reply:
It is GC'd, but leaving threads lying around is not good practice as they are an OS resource and hold a chunk of memory too.  A bug such as this can result in requests failing due to insufficient resources, especially in something like pull queries that can see high call volumes.

Reply:
PullQueryExecutor was created once per application before the refactor.  @vpapavas did benchmark this with the new physical plan code, but it must not have shown up only doing high qps for ~30 minutes at a time.  Well, good to catch before hitting the soak cluster.

Reply:
If it's not too big of a change for this PR, it would be great to make the threadpool one per application.  Otherwise we can do it after this PR.  We should make sure we do that before 0.15 regardless.

Reply:
https://github.com/confluentinc/ksql/issues/6666